File: api/load-vault.py
import json
import os
import io
import zipfile
import xml.etree.ElementTree as ET
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
from google.oauth2.service_account import Credentials
from googleapiclient.discovery import build
import requests
import gzip
import base64

# Vault folder ID
VAULT_FOLDER_ID = "1LAkbqjN7g-HJV9BRWV-AsmMpY1JzJiIM"

def extract_text_from_docx(docx_data):
    """Extract text content from DOCX file data"""
    try:
        with zipfile.ZipFile(io.BytesIO(docx_data), 'r') as zip_file:
            document_xml = zip_file.read('word/document.xml')
            root = ET.fromstring(document_xml)
            namespaces = {
                'w': 'http://schemas.openxmlformats.org/wordprocessingml/2006/main'
            }
            text_elements = root.findall('.//w:t', namespaces)
            text_content = []
            for elem in text_elements:
                if elem.text:
                    text_content.append(elem.text)
            extracted_text = ' '.join(text_content)
            lines = extracted_text.split('\n')
            cleaned_lines = [line.strip() for line in lines if line.strip()]
            return '\n'.join(cleaned_lines)
    except Exception as e:
        return f"[DOCX text extraction failed: {str(e)}]"

def get_google_drive_service():
    """Initialize Google Drive service with credentials"""
    try:
        creds_json = os.environ.get('GOOGLE_CREDENTIALS_JSON')
        project_id = os.environ.get('GOOGLE_PROJECT_ID')
        project_number = os.environ.get('GOOGLE_PROJECT_NUMBER')
        
        if not creds_json:
            raise Exception("GOOGLE_CREDENTIALS_JSON environment variable not found")
        if not project_id:
            raise Exception("GOOGLE_PROJECT_ID environment variable not found")
        
        print(f"Using Project ID: {project_id}")
        print(f"Using Project Number: {project_number}")
        
        creds_info = json.loads(creds_json)
        creds_info['project_id'] = project_id
        if project_number:
            creds_info['project_number'] = project_number
            
        creds = Credentials.from_service_account_info(
            creds_info,
            scopes=['https://www.googleapis.com/auth/drive.readonly']
        )
        
        return build('drive', 'v3', credentials=creds)
    except Exception as e:
        raise Exception(f"Google Drive authentication failed: {str(e)}")

def store_vault_in_kv(vault_data):
    """Store vault data in Vercel KV using v2 key"""
    try:
        kv_url = os.environ.get('KV_REST_API_URL')
        kv_token = os.environ.get('KV_REST_API_TOKEN')
        
        if not kv_url or not kv_token:
            print("⚠️ KV environment variables not found")
            return False
            
        headers = {
            'Authorization': f'Bearer {kv_token}',
            'Content-Type': 'application/json'
        }
        
        # Convert to JSON and check size
        vault_json = json.dumps(vault_data)
        vault_size = len(vault_json)
        
        print(f"📊 Vault data size: {vault_size} bytes")
        
        # Compress if over 80KB
        if vault_size > 80000:
            print("🗜️ Compressing vault data...")
            compressed = gzip.compress(vault_json.encode('utf-8'))
            encoded = base64.b64encode(compressed).decode('utf-8')
            
            # Store compressed with metadata
            compressed_data = {
                "compressed": True,
                "data": encoded,
                "original_size": vault_size,
                "compressed_size": len(encoded)
            }
            
            response = requests.post(
                f'{kv_url}/set/sitemonkeys_vault_v2',
                headers=headers,
                data=json.dumps(compressed_data),
                timeout=30
            )
            
            print(f"📦 Compressed: {vault_size} → {len(encoded)} bytes")
        else:
            # Store uncompressed
            response = requests.post(
                f'{kv_url}/set/sitemonkeys_vault_v2',
                headers=headers,
                data=vault_json,
                timeout=30
            )
        
        print(f"🔑 KV Storage URL: {kv_url}/set/sitemonkeys_vault_v2")
        print(f"📤 KV Storage response status: {response.status_code}")
        print(f"📤 KV Storage response: {response.text}")
        
        if response.status_code == 200:
            print("✅ Vault data stored in KV successfully")
            
            # Store master index separately for diagnostics
            master_index = {
                "folders_loaded": vault_data.get("folders_loaded", []),
                "total_files": vault_data.get("total_files", 0),
                "tokens": vault_data.get("tokens", 0),
                "estimated_cost": vault_data.get("estimated_cost", "$0.00"),
                "last_updated": vault_data.get("last_updated", "unknown"),
                "vault_status": "operational",
                "storage_method": "compressed" if vault_size > 80000 else "direct"
            }
            
            index_response = requests.post(
                f'{kv_url}/set/sitemonkeys_vault_v2_index',
                headers=headers,
                data=json.dumps(master_index),
                timeout=30
            )
            
            print(f"📋 Index storage status: {index_response.status_code}")
            return True
        else:
            print(f"❌ KV storage failed: {response.status_code} - {response.text}")
            return False
            
    except Exception as e:
        print(f"❌ KV storage error: {str(e)}")
        import traceback
        print(f"Full traceback: {traceback.format_exc()}")
        return False

def get_vault_from_kv():
    """Retrieve vault data from Vercel KV using v2 key"""
    try:
        kv_url = os.environ.get('KV_REST_API_URL')
        kv_token = os.environ.get('KV_REST_API_TOKEN')
        
        if not kv_url or not kv_token:
            print("⚠️ KV environment variables not found for retrieval")
            return None
            
        headers = {
            'Authorization': f'Bearer {kv_token}',
        }
        
        response = requests.get(
            f'{kv_url}/get/sitemonkeys_vault_v2',
            headers=headers,
            timeout=30
        )
        
        print(f"🔍 KV Retrieval URL: {kv_url}/get/sitemonkeys_vault_v2")
        print(f"📥 KV Retrieval response status: {response.status_code}")
        print(f"📥 KV Retrieval response length: {len(response.text) if response.text else 0}")
        
        if response.status_code == 200:
            try:
                response_text = response.text.strip()
                if response_text and response_text != 'null':
                    # Try to parse as JSON
                    data = json.loads(response_text)
                    
                    # Check if it's compressed
                    if isinstance(data, dict) and data.get("compressed"):
                        print("🗜️ Decompressing vault data...")
                        compressed_data = base64.b64decode(data["data"])
                        decompressed = gzip.decompress(compressed_data)
                        result = json.loads(decompressed.decode('utf-8'))
                        print(f"📦 Decompressed: {data.get('compressed_size', 0)} → {data.get('original_size', 0)} bytes")
                    else:
                        result = data
                    
                    print(f"✅ Retrieved vault data from KV: {len(str(result))} characters")
                    return result
                else:
                    print("⚠️ No vault data found in KV (null response)")
                    return None
                    
            except json.JSONDecodeError as e:
                print(f"❌ Failed to parse KV response as JSON: {e}")
                print(f"Raw response: {response.text[:500]}...")
                return None
        else:
            print(f"❌ KV retrieval failed: {response.status_code}")
            print(f"Error response: {response.text}")
            return None
            
    except Exception as e:
        print(f"❌ KV retrieval error: {str(e)}")
        import traceback
        print(f"Full traceback: {traceback.format_exc()}")
        return None

def load_vault_content():
    """Load all content from SiteMonkeys vault folders"""
    vault_content = "=== SITEMONKEYS BUSINESS VALIDATION VAULT ===\n\n"
    loaded_folders = []
    total_files = 0
    
    try:
        service = get_google_drive_service()
        
        # Get subfolders in vault
        query = f"'{VAULT_FOLDER_ID}' in parents and mimeType='application/vnd.google-apps.folder'"
        folders_result = service.files().list(q=query, fields="files(id, name)", pageSize=50).execute()
        folders = folders_result.get('files', [])
        
        vault_content += f"\n=== LIVE VAULT FOLDERS LOADED ({len(loaded_folders)} folders) ===\n"
        
        # ZERO-FAILURE: Filter to only the 3 required folders
        target_folders = ['00_EnforcementShell', '01_Core_Directives', 'VAULT_MEMORY_FILES']
        
        for folder in folders:
            folder_name = folder['name']
            
            # Skip folders not in our target list
            if folder_name not in target_folders:
                print(f"⏭️ Skipping folder: {folder_name} (not in target list)")
                continue
                
            loaded_folders.append(folder_name)
            vault_content += f"\n--- FOLDER: {folder_name} ---\n"
            
            # Get files in each folder
            file_query = f"'{folder['id']}' in parents"
            files_result = service.files().list(
                q=file_query, 
                fields="files(id, name, mimeType, size)",
                pageSize=100
            ).execute()
            files = files_result.get('files', [])
            
            total_files += len(files)
            
            for file in files:
                try:
                    file_content = ""
                    file_mime = file.get('mimeType', '')
                    file_name = file['name']
                    
                    print(f"Processing file: {file_name} (type: {file_mime})")
                    
                    if 'text/plain' in file_mime or file_name.endswith('.txt'):
                        try:
                            file_data = service.files().get_media(fileId=file['id']).execute()
                            file_content = file_data.decode('utf-8')
                            print(f"✅ Text file loaded: {len(file_content)} characters")
                        except Exception as e:
                            file_content = f"[ERROR reading text file: {str(e)}]"
                            print(f"❌ Text file error: {str(e)}")
                    
                    elif file_mime == 'application/vnd.google-apps.document':
                        try:
                            export_data = service.files().export(
                                fileId=file['id'], 
                                mimeType='text/plain'
                            ).execute()
                            file_content = export_data.decode('utf-8')
                            print(f"✅ Google Doc exported: {len(file_content)} characters")
                        except Exception as e:
                            file_content = f"[Google Doc - Export failed: {str(e)}]"
                            print(f"❌ Google Doc error: {str(e)}")
                    
                    elif 'application/vnd.openxmlformats-officedocument.wordprocessingml.document' in file_mime or file_name.endswith('.docx'):
                        try:
                            print(f"🔄 Downloading DOCX file: {file_name}")
                            file_data = service.files().get_media(fileId=file['id']).execute()
                            print(f"📥 Downloaded {len(file_data)} bytes")
                            
                            file_content = extract_text_from_docx(file_data)
                            
                            if file_content.startswith('[DOCX text extraction failed'):
                                print(f"❌ DOCX extraction failed: {file_name}")
                            else:
                                print(f"✅ DOCX text extracted: {len(file_content)} characters")
                                
                        except Exception as e:
                            file_content = f"[DOCX file - Access failed: {str(e)}]"
                            print(f"❌ DOCX access error: {str(e)}")
                    
                    elif file_mime == 'application/vnd.google-apps.folder':
                        continue
                    
                    else:
                        file_size = file.get('size', 'Unknown')
                        file_content = f"[File type: {file_mime} - Size: {file_size} bytes - Skipped unsupported format]"
                        print(f"⏭️ Skipped unsupported file: {file_name}")
                    
                    vault_content += f"\n=== {file_name} ===\n{file_content}\n"
                    
                except Exception as file_error:
                    vault_content += f"\n=== {file['name']} ===\n[ERROR loading file: {str(file_error)}]\n"
                    print(f"❌ File processing error: {file['name']} - {str(file_error)}")
        
        vault_content += f"\n=== VAULT SUMMARY ===\nFolders: {len(loaded_folders)}\nFiles processed: {total_files}\n"
        
        print(f"✅ Vault loaded successfully: {len(folders)} folders, {total_files} files")
        print(f"📊 Total vault content: {len(vault_content)} characters")
                      
    except Exception as drive_error:
        print(f"❌ Google Drive error: {str(drive_error)}")
        vault_content += f"\n[Google Drive connection error: {str(drive_error)}]\n"
        loaded_folders = ["Error - Could not connect to Google Drive"]
        total_files = 0
    
    return vault_content, loaded_folders, total_files

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
        
        try:
            parsed_url = urlparse(self.path)
            query_params = parse_qs(parsed_url.query)
            is_refresh = 'refresh' in query_params and query_params['refresh'][0] == 'true'
            
            if is_refresh:
                print("🔄 Refresh requested - loading fresh vault data...")
                
                vault_memory, loaded_folders, total_files = load_vault_content()
                
                token_count = len(vault_memory) // 4
                estimated_cost = (token_count * 0.002) / 1000
                
                vault_data = {
                    "vault_content": vault_memory,
                    "tokens": token_count,
                    "estimated_cost": f"${estimated_cost:.4f}",
                    "folders_loaded": loaded_folders,
                    "total_files": total_files,
                    "last_updated": "refresh_requested",
                    "vault_status": "operational"
                }
                
                kv_stored = store_vault_in_kv(vault_data)
                
                print(f"📊 Vault refresh complete: {token_count} tokens, {len(loaded_folders)} folders")
                
                response = {
                    "status": "refreshed",
                    "vault_content": vault_memory,
                    "tokens": token_count,
                    "estimated_cost": f"${estimated_cost:.4f}",
                    "folders_loaded": loaded_folders,
                    "total_files": total_files,
                    "kv_stored": kv_stored,
                    "message": f"Vault refreshed: {len(loaded_folders)} folders, {total_files} files",
                    "vault_status": "operational"
                }
                
            else:
                print("📖 Checking for cached vault data...")
                
                cached_vault = get_vault_from_kv()
                
                if cached_vault and isinstance(cached_vault, dict) and cached_vault.get("vault_content"):
                    print("✅ Found valid cached vault data in KV")
                    response = {
                        "status": "success",
                        "vault_content": cached_vault.get("vault_content", ""),
                        "tokens": cached_vault.get("tokens", 0),
                        "estimated_cost": cached_vault.get("estimated_cost", "$0.00"),
                        "folders_loaded": cached_vault.get("folders_loaded", []),
                        "total_files": cached_vault.get("total_files", 0),
                        "vault_status": cached_vault.get("vault_status", "operational"),
                        "message": "Using cached vault data from KV"
                    }
                else:
                    print("⚠️ No valid cached vault data found")
                    response = {
                        "status": "success",
                        "needs_refresh": True,
                        "vault_content": "",
                        "tokens": 0,
                        "estimated_cost": "$0.00",
                        "folders_loaded": [],
                        "total_files": 0,
                        "vault_status": "needs_refresh",
                        "message": "No vault data found - please refresh"
                    }
            
            self.wfile.write(json.dumps(response).encode())
            
        except Exception as e:
            print(f"❌ Vault operation failed: {str(e)}")
            error_response = {
                "status": "error",
                "error": str(e),
                "vault_status": "error",
                "message": "Vault operation failed - check configuration"
            }
            self.wfile.write(json.dumps(error_response).encode())
    
    def do_POST(self):
        self.do_GET()
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

File: api/load-vault.js
const express = require('express');
const cors = require('cors');
const { google } = require('googleapis');
const axios = require('axios');
const JSZip = require('jszip');
const xml2js = require('xml2js');
const zlib = require('zlib');
const { promisify } = require('util');

const gzip = promisify(zlib.gzip);
const gunzip = promisify(zlib.gunzip);

// Vault folder ID
const VAULT_FOLDER_ID = "1LAkbqjN7g-HJV9BRWV-AsmMpY1JzJiIM";

/**
 * Extract text content from DOCX file data
 */
async function extractTextFromDocx(docxBuffer) {
    try {
        const zip = new JSZip();
        const zipContent = await zip.loadAsync(docxBuffer);
        
        const documentXml = await zipContent.file('word/document.xml').async('string');
        
        const parser = new xml2js.Parser();
        const result = await parser.parseStringPromise(documentXml);
        
        const textElements = [];
        
        function extractText(node) {
            if (node && typeof node === 'object') {
                if (node['w:t'] && Array.isArray(node['w:t'])) {
                    node['w:t'].forEach(textNode => {
                        if (textNode && textNode._) {
                            textElements.push(textNode._);
                        } else if (typeof textNode === 'string') {
                            textElements.push(textNode);
                        }
                    });
                }
                
                Object.keys(node).forEach(key => {
                    if (Array.isArray(node[key])) {
                        node[key].forEach(child => extractText(child));
                    } else if (typeof node[key] === 'object') {
                        extractText(node[key]);
                    }
                });
            }
        }
        
        extractText(result);
        
        const extractedText = textElements.join(' ');
        const lines = extractedText.split('\n');
        const cleanedLines = lines.filter(line => line.trim()).map(line => line.trim());
        
        return cleanedLines.join('\n');
    } catch (error) {
        return `[DOCX text extraction failed: ${error.message}]`;
    }
}

/**
 * Initialize Google Drive service with credentials
 */
function getGoogleDriveService() {
    try {
        const credsJson = process.env.GOOGLE_CREDENTIALS_JSON;
        const projectId = process.env.GOOGLE_PROJECT_ID;
        const projectNumber = process.env.GOOGLE_PROJECT_NUMBER;
        
        if (!credsJson) {
            throw new Error("GOOGLE_CREDENTIALS_JSON environment variable not found");
        }
        if (!projectId) {
            throw new Error("GOOGLE_PROJECT_ID environment variable not found");
        }
        
        console.log(`Using Project ID: ${projectId}`);
        console.log(`Using Project Number: ${projectNumber}`);
        
        const credsInfo = JSON.parse(credsJson);
        credsInfo.project_id = projectId;
        if (projectNumber) {
            credsInfo.project_number = projectNumber;
        }
        
        const auth = new google.auth.GoogleAuth({
            credentials: credsInfo,
            scopes: ['https://www.googleapis.com/auth/drive.readonly']
        });
        
        return google.drive({ version: 'v3', auth });
    } catch (error) {
        throw new Error(`Google Drive authentication failed: ${error.message}`);
    }
}

/**
 * Store vault data in Railway KV using v2 key
 */
async function storeVaultInKv(vaultData) {
    try {
        const kvUrl = process.env.KV_REST_API_URL;
        const kvToken = process.env.KV_REST_API_TOKEN;
        
        if (!kvUrl || !kvToken) {
            console.log("⚠️ KV environment variables not found");
            return false;
        }
        
        const headers = {
            'Authorization': `Bearer ${kvToken}`,
            'Content-Type': 'application/json'
        };
        
        // Convert to JSON and check size
        const vaultJson = JSON.stringify(vaultData);
        const vaultSize = Buffer.byteLength(vaultJson, 'utf8');
        
        console.log(`📊 Vault data size: ${vaultSize} bytes`);
        
        let requestData;
        let storageMethod = 'direct';
        
        // Compress if over 80KB
        if (vaultSize > 80000) {
            console.log("🗜️ Compressing vault data...");
            const compressed = await gzip(Buffer.from(vaultJson, 'utf8'));
            const encoded = compressed.toString('base64');
            
            // Store compressed with metadata
            requestData = {
                compressed: true,
                data: encoded,
                original_size: vaultSize,
                compressed_size: encoded.length
            };
            
            storageMethod = 'compressed';
            console.log(`📦 Compressed: ${vaultSize} → ${encoded.length} bytes`);
        } else {
            requestData = vaultData;
        }
        
        const response = await axios.post(
            `${kvUrl}/set/sitemonkeys_vault_v2`,
            requestData,
            { 
                headers,
                timeout: 30000
            }
        );
        
        console.log(`🔑 KV Storage URL: ${kvUrl}/set/sitemonkeys_vault_v2`);
        console.log(`📤 KV Storage response status: ${response.status}`);
        console.log(`📤 KV Storage response: ${response.data}`);
        
        if (response.status === 200) {
            console.log("✅ Vault data stored in KV successfully");
            
            // Store master index separately for diagnostics
            const masterIndex = {
                folders_loaded: vaultData.folders_loaded || [],
                total_files: vaultData.total_files || 0,
                tokens: vaultData.tokens || 0,
                estimated_cost: vaultData.estimated_cost || "$0.00",
                last_updated: vaultData.last_updated || "unknown",
                vault_status: "operational",
                storage_method: storageMethod
            };
            
            const indexResponse = await axios.post(
                `${kvUrl}/set/sitemonkeys_vault_v2_index`,
                masterIndex,
                { 
                    headers,
                    timeout: 30000
                }
            );
            
            console.log(`📋 Index storage status: ${indexResponse.status}`);
            return true;
        } else {
            console.log(`❌ KV storage failed: ${response.status} - ${response.data}`);
            return false;
        }
        
    } catch (error) {
        console.log(`❌ KV storage error: ${error.message}`);
        console.log(`Full error: ${error.stack}`);
        return false;
    }
}

/**
 * Retrieve vault data from Railway KV using v2 key
 */
async function getVaultFromKv() {
    try {
        const kvUrl = process.env.KV_REST_API_URL;
        const kvToken = process.env.KV_REST_API_TOKEN;
        
        if (!kvUrl || !kvToken) {
            console.log("⚠️ KV environment variables not found for retrieval");
            return null;
        }
        
        const headers = {
            'Authorization': `Bearer ${kvToken}`,
        };
        
        const response = await axios.get(
            `${kvUrl}/get/sitemonkeys_vault_v2`,
            { 
                headers,
                timeout: 30000
            }
        );
        
        console.log(`🔍 KV Retrieval URL: ${kvUrl}/get/sitemonkeys_vault_v2`);
        console.log(`📥 KV Retrieval response status: ${response.status}`);
        console.log(`📥 KV Retrieval response length: ${response.data ? JSON.stringify(response.data).length : 0}`);
        
        if (response.status === 200) {
            try {
                const data = response.data;
                
                if (data === null || data === undefined) {
                    console.log("⚠️ No vault data found in KV (null response)");
                    return null;
                }
                
                // Check if it's compressed
                if (typeof data === 'object' && data.compressed) {
                    console.log("🗜️ Decompressing vault data...");
                    const compressedBuffer = Buffer.from(data.data, 'base64');
                    const decompressed = await gunzip(compressedBuffer);
                    const result = JSON.parse(decompressed.toString('utf8'));
                    console.log(`📦 Decompressed: ${data.compressed_size || 0} → ${data.original_size || 0} bytes`);
                    return result;
                } else {
                    console.log(`✅ Retrieved vault data from KV: ${JSON.stringify(data).length} characters`);
                    return data;
                }
                
            } catch (parseError) {
                console.log(`❌ Failed to parse KV response: ${parseError.message}`);
                console.log(`Raw response: ${JSON.stringify(response.data).substring(0, 500)}...`);
                return null;
            }
        } else {
            console.log(`❌ KV retrieval failed: ${response.status}`);
            console.log(`Error response: ${response.data}`);
            return null;
        }
        
    } catch (error) {
        console.log(`❌ KV retrieval error: ${error.message}`);
        console.log(`Full error: ${error.stack}`);
        return null;
    }
}

/**
 * Load all content from SiteMonkeys vault folders
 */
async function loadVaultContent() {
    let vaultContent = "=== SITEMONKEYS BUSINESS VALIDATION VAULT ===\n\n";
    let loadedFolders = [];
    let totalFiles = 0;
    
    try {
        const drive = getGoogleDriveService();
        
        // Get subfolders in vault
        const query = `'${VAULT_FOLDER_ID}' in parents and mimeType='application/vnd.google-apps.folder'`;
        const foldersResult = await drive.files.list({
            q: query,
            fields: "files(id, name)",
            pageSize: 50
        });
        
        const folders = foldersResult.data.files || [];
        
        vaultContent += `\n=== LIVE VAULT FOLDERS LOADED (${loadedFolders.length} folders) ===\n`;
        
        // ZERO-FAILURE: Filter to only the 3 required folders
        const targetFolders = ['00_EnforcementShell', '01_Core_Directives', 'VAULT_MEMORY_FILES'];
        
        for (const folder of folders) {
            const folderName = folder.name;
            
            // Skip folders not in our target list
            if (!targetFolders.includes(folderName)) {
                console.log(`⏭️ Skipping folder: ${folderName} (not in target list)`);
                continue;
            }
            
            loadedFolders.push(folderName);
            vaultContent += `\n--- FOLDER: ${folderName} ---\n`;
            
            // Get files in each folder
            const fileQuery = `'${folder.id}' in parents`;
            const filesResult = await drive.files.list({
                q: fileQuery,
                fields: "files(id, name, mimeType, size)",
                pageSize: 100
            });
            
            const files = filesResult.data.files || [];
            totalFiles += files.length;
            
            for (const file of files) {
                try {
                    let fileContent = "";
                    const fileMime = file.mimeType || '';
                    const fileName = file.name;
                    
                    console.log(`Processing file: ${fileName} (type: ${fileMime})`);
                    
                    if (fileMime.includes('text/plain') || fileName.endsWith('.txt')) {
                        try {
                            const response = await drive.files.get({
                                fileId: file.id,
                                alt: 'media'
                            }, { responseType: 'arraybuffer' });
                            
                            fileContent = Buffer.from(response.data).toString('utf-8');
                            console.log(`✅ Text file loaded: ${fileContent.length} characters`);
                        } catch (error) {
                            fileContent = `[ERROR reading text file: ${error.message}]`;
                            console.log(`❌ Text file error: ${error.message}`);
                        }
                    }
                    else if (fileMime === 'application/vnd.google-apps.document') {
                        try {
                            const response = await drive.files.export({
                                fileId: file.id,
                                mimeType: 'text/plain'
                            });
                            
                            fileContent = response.data;
                            console.log(`✅ Google Doc exported: ${fileContent.length} characters`);
                        } catch (error) {
                            fileContent = `[Google Doc - Export failed: ${error.message}]`;
                            console.log(`❌ Google Doc error: ${error.message}`);
                        }
                    }
                    else if (fileMime.includes('application/vnd.openxmlformats-officedocument.wordprocessingml.document') || fileName.endsWith('.docx')) {
                        try {
                            console.log(`🔄 Downloading DOCX file: ${fileName}`);
                            const response = await drive.files.get({
                                fileId: file.id,
                                alt: 'media'
                            }, { responseType: 'arraybuffer' });
                            
                            const fileData = Buffer.from(response.data);
                            console.log(`📥 Downloaded ${fileData.length} bytes`);
                            
                            fileContent = await extractTextFromDocx(fileData);
                            
                            if (fileContent.startsWith('[DOCX text extraction failed')) {
                                console.log(`❌ DOCX extraction failed: ${fileName}`);
                            } else {
                                console.log(`✅ DOCX text extracted: ${fileContent.length} characters`);
                            }
                            
                        } catch (error) {
                            fileContent = `[DOCX file - Access failed: ${error.message}]`;
                            console.log(`❌ DOCX access error: ${error.message}`);
                        }
                    }
                    else if (fileMime === 'application/vnd.google-apps.folder') {
                        continue;
                    }
                    else {
                        const fileSize = file.size || 'Unknown';
                        fileContent = `[File type: ${fileMime} - Size: ${fileSize} bytes - Skipped unsupported format]`;
                        console.log(`⏭️ Skipped unsupported file: ${fileName}`);
                    }
                    
                    vaultContent += `\n=== ${fileName} ===\n${fileContent}\n`;
                    
                } catch (fileError) {
                    vaultContent += `\n=== ${file.name} ===\n[ERROR loading file: ${fileError.message}]\n`;
                    console.log(`❌ File processing error: ${file.name} - ${fileError.message}`);
                }
            }
        }
        
        vaultContent += `\n=== VAULT SUMMARY ===\nFolders: ${loadedFolders.length}\nFiles processed: ${totalFiles}\n`;
        
        console.log(`✅ Vault loaded successfully: ${folders.length} folders, ${totalFiles} files`);
        console.log(`📊 Total vault content: ${vaultContent.length} characters`);
        
    } catch (driveError) {
        console.log(`❌ Google Drive error: ${driveError.message}`);
        vaultContent += `\n[Google Drive connection error: ${driveError.message}]\n`;
        loadedFolders = ["Error - Could not connect to Google Drive"];
        totalFiles = 0;
    }
    
    return { vaultContent, loadedFolders, totalFiles };
}

// Express app setup
const app = express();

// Middleware
app.use(cors({
    origin: '*',
    methods: ['GET', 'POST', 'OPTIONS'],
    allowedHeaders: ['Content-Type']
}));
app.use(express.json());

// Main vault endpoint - handles both GET and POST
app.all('/api/load-vault', async (req, res) => {
    try {
        const isRefresh = req.query.refresh === 'true';
        
        if (isRefresh) {
            console.log("🔄 Refresh requested - loading fresh vault data...");
            
            const { vaultContent: vaultMemory, loadedFolders, totalFiles } = await loadVaultContent();
            
            const tokenCount = Math.floor(vaultMemory.length / 4);
            const estimatedCost = (tokenCount * 0.002) / 1000;
            
            const vaultData = {
                vault_content: vaultMemory,
                tokens: tokenCount,
                estimated_cost: `$${estimatedCost.toFixed(4)}`,
                folders_loaded: loadedFolders,
                total_files: totalFiles,
                last_updated: "refresh_requested",
                vault_status: "operational"
            };
            
            const kvStored = await storeVaultInKv(vaultData);
            
            console.log(`📊 Vault refresh complete: ${tokenCount} tokens, ${loadedFolders.length} folders`);
            
            const response = {
                status: "refreshed",
                vault_content: vaultMemory,
                tokens: tokenCount,
                estimated_cost: `$${estimatedCost.toFixed(4)}`,
                folders_loaded: loadedFolders,
                total_files: totalFiles,
                kv_stored: kvStored,
                message: `Vault refreshed: ${loadedFolders.length} folders, ${totalFiles} files`,
                vault_status: "operational"
            };
            
            res.json(response);
            
        } else {
            console.log("📖 Checking for cached vault data...");
            
            const cachedVault = await getVaultFromKv();
            
            if (cachedVault && typeof cachedVault === 'object' && cachedVault.vault_content) {
                console.log("✅ Found valid cached vault data in KV");
                const response = {
                    status: "success",
                    vault_content: cachedVault.vault_content || "",
                    tokens: cachedVault.tokens || 0,
                    estimated_cost: cachedVault.estimated_cost || "$0.00",
                    folders_loaded: cachedVault.folders_loaded || [],
                    total_files: cachedVault.total_files || 0,
                    vault_status: cachedVault.vault_status || "operational",
                    message: "Using cached vault data from KV"
                };
                res.json(response);
            } else {
                console.log("⚠️ No valid cached vault data found");
                const response = {
                    status: "success",
                    needs_refresh: true,
                    vault_content: "",
                    tokens: 0,
                    estimated_cost: "$0.00",
                    folders_loaded: [],
                    total_files: 0,
                    vault_status: "needs_refresh",
                    message: "No vault data found - please refresh"
                };
                res.json(response);
            }
        }
        
    } catch (error) {
        console.log(`❌ Vault operation failed: ${error.message}`);
        const errorResponse = {
            status: "error",
            error: error.message,
            vault_status: "error",
            message: "Vault operation failed - check configuration"
        };
        res.json(errorResponse);
    }
});

// Health check endpoint
app.get('/', (req, res) => {
    res.json({ 
        status: 'ok', 
        service: 'sitemonkeys-vault-loader',
        message: 'Vault loader service is running',
        endpoints: ['/api/load-vault', '/health']
    });
});

app.get('/health', (req, res) => {
    res.json({ status: 'ok', service: 'vault-loader' });
});

// Start server
const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`🚀 SiteMonkeys Vault loader service running on port ${PORT}`);
    console.log(`📍 Vault endpoint: http://localhost:${PORT}/api/load-vault`);
});
